### PR TITLE
 ✨feat(rust/tree): add binary_search_tree

### PR DESCRIPTION
**Pull request by xBLACKICEx**
If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).
image

### CODE REVIEW

**Comment by krahets** [comment]
To be reviewed by @sjinzh

**Review by sjinzh — APPROVED** [review]
@krahets It's oK.

File: codes/rust/chapter_tree/binary_search_tree.rs [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+    /* 查找节点 */
+    let node = bst.search(7).unwrap();
+    println!(
+        "\n查找到的节点对象为: {:p} 节点值 = {}\n",
+        node.as_ref().as_ptr(),
+        node.borrow().val
+    );
+
+    // /* 插入节点 */
```
Comment (krahets) [inline codes/rust/chapter_tree/binary_search_tree.rs:186]: Fomat

File: codes/rust/chapter_tree/binary_search_tree.rs [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                // 递归删除节点 next
+                self.remove(tmp.borrow().val);
+                // 将 nex 的值复制给 cur
+                cur.borrow_mut().val = tmp.borrow().val;
+            }
+        }
+    }
+
+    /* 获取中序遍历中的下一个节点（仅适用于 root 有左子节点的情况） */
+    fn get_next_node(mut root: Option<TreeNodeRc>) -> Option<TreeNodeRc> {
```
Comment (krahets) [inline codes/rust/chapter_tree/binary_search_tree.rs:157]: The latest code has removed this function. Please check it.

File: codes/rust/chapter_tree/binary_search_tree.rs [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        } else {
+            Self {
+                root: Some(Self::build_tree(&num)), // 构建二叉搜索树
+            }
+        }
+    }
+
+    /* 获取二叉树根节点 */
+    pub fn get_root(&self) -> Option<TreeNodeRc> {
+        self.root.clone() // RC 克隆
```
Comment (krahets) [inline codes/rust/chapter_tree/binary_search_tree.rs:29]: Why clone the root node?

File: codes/rust/chapter_tree/binary_search_tree.rs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    /* 插入节点 */
+    pub fn insert(&mut self, num: i32) -> Option<TreeNodeRc> {
+        // 若树为空，直接提前返回
+        if self.root.is_none() {
+            return None;
+        }
+
+        let mut cur = self.root.clone();
```
Comment (krahets) [inline codes/rust/chapter_tree/binary_search_tree.rs:77]: I'm confused about `clone()` . Is it clone a new object of the `root` node? Is it necessary?

**Review by krahets — COMMENTED** [review]
Hi @xBLACKICEx  Please address the comment, thanks!

**Review by krahets — APPROVED** [review]
Thanks for the PR!

**Pull request merged**